File: cli/lib/scanner.js
import fs from 'node:fs/promises';
import path from 'node:path';
import { ARTIFACT_CATEGORIES, MAX_DEPTH, getSkipDirs } from './constants.js';

/**
 * Build lookup structures from selected category IDs.
 */
function buildTargetMap(categoryIds) {
  const targetMap = new Map(); // dirName -> [categoryDef, ...]
  const allTargetDirs = new Set();

  for (const id of categoryIds) {
    const def = ARTIFACT_CATEGORIES[id];
    if (!def || def.isSystem) continue;

    for (const dirName of def.targetDirs) {
      allTargetDirs.add(dirName);
      if (!targetMap.has(dirName)) {
        targetMap.set(dirName, []);
      }
      targetMap.get(dirName).push(def);
    }
  }

  return { targetMap, allTargetDirs };
}

/**
 * Check if a directory name matches an artifact definition, including sibling file detection.
 */
async function matchDefinition(dirName, parentPath, targetMap) {
  const defs = targetMap.get(dirName);
  if (!defs) return null;

  for (const def of defs) {
    if (def.siblingFiles.length === 0) {
      return def;
    }
    for (const sibling of def.siblingFiles) {
      try {
        await fs.access(path.join(parentPath, sibling));
        return def;
      } catch {
        // sibling not found, try next
      }
    }
  }
  return null;
}

/**
 * Scan a directory tree for artifact directories matching selected categories.
 * Uses iterative DFS with fs.opendir for memory efficiency.
 *
 * @param {string} root - Root directory to scan
 * @param {string[]} categoryIds - Array of category IDs to scan for
 * @param {object} options
 * @param {function} [options.onProgress] - Called with current directory path
 * @param {function} [options.onFound] - Called with (path, categoryId, totalFound)
 * @param {boolean} [options.includeHidden] - Also scan hidden dirs not in skip list
 * @param {object} [options.signal] - Object with { cancelled: boolean } to abort
 * @returns {Promise<Array<{path: string, categoryId: string}>>}
 */
export async function scanForArtifacts(root, categoryIds, options = {}) {
  const { onProgress, onFound, includeHidden = false, signal } = options;
  const results = [];

  const { targetMap, allTargetDirs } = buildTargetMap(categoryIds);
  if (allTargetDirs.size === 0) return results;

  // Build skip dirs, removing any that are scan targets for selected categories
  const skipDirs = getSkipDirs(categoryIds);

  const stack = [[path.resolve(root), 0]];

  while (stack.length > 0) {
    if (signal?.cancelled) return results;

    const [dirPath, depth] = stack.pop();
    if (depth > MAX_DEPTH) continue;

    if (onProgress) {
      onProgress(dirPath);
    }

    let dir;
    try {
      dir = await fs.opendir(dirPath);
    } catch {
      continue;
    }

    try {
      for await (const entry of dir) {
        const name = entry.name;

        // Check if this is a target directory
        if (allTargetDirs.has(name)) {
          const fullPath = path.join(dirPath, name);
          try {
            const stat = await fs.lstat(fullPath);
            if (stat.isDirectory() && !stat.isSymbolicLink()) {
              const def = await matchDefinition(name, dirPath, targetMap);
              if (def) {
                results.push({ path: fullPath, categoryId: def.id });
                if (onFound) onFound(fullPath, def.id, results.length);
              }
            }
          } catch {
            // Can't stat -- skip
          }
          continue; // Don't recurse into matched dirs
        }

        if (!entry.isDirectory()) continue;
        if (entry.isSymbolicLink()) continue;
        if (skipDirs.has(name)) continue;
        if (!includeHidden && name.startsWith('.') && !allTargetDirs.has(name)) continue;

        stack.push([path.join(dirPath, name), depth + 1]);
      }
    } catch {
      // Error reading directory entries
    }
  }

  return results;
}

/**
 * Check system-level artifact paths.
 * @param {string[]} categoryIds - Array of category IDs to check
 * @returns {Promise<Array<{path: string, categoryId: string}>>}
 */
export async function checkSystemArtifacts(categoryIds) {
  const results = [];

  for (const id of categoryIds) {
    const def = ARTIFACT_CATEGORIES[id];
    if (!def || !def.isSystem) continue;

    try {
      const stat = await fs.stat(def.systemPath);
      if (!stat.isDirectory()) continue;
    } catch {
      continue; // Path doesn't exist or not accessible
    }

    if (def.expandSubdirs) {
      try {
        const entries = await fs.readdir(def.systemPath, { withFileTypes: true });
        for (const entry of entries) {
          if (entry.isDirectory()) {
            results.push({
              path: path.join(def.systemPath, entry.name),
              categoryId: id,
            });
          }
        }
      } catch {
        // If we can't list subdirs, add the whole directory
        results.push({ path: def.systemPath, categoryId: id });
      }
    } else {
      results.push({ path: def.systemPath, categoryId: id });
    }
  }

  return results;
}

File: Sources/Scanner.swift
import Foundation

actor Scanner {
    private static let baseSkipDirs: Set<String> = [
        "Library", ".Trash", "Applications", "Pictures", "Movies", "Music",
        ".Spotlight-V100", ".fseventsd",
        ".npm", ".cache", ".yarn", ".pnpm-store", ".bun",
        ".git", ".docker", ".ollama", ".conda", ".anaconda",
        ".cocoapods", ".cargo", ".rustup",
        ".oh-my-zsh", ".zsh_sessions",
        ".vscode", ".cursor", ".windsurf",
        ".local", ".config", "Wallpapers",
    ]

    private var cancelled = false

    func cancel() {
        cancelled = true
    }

    /// Scan a directory tree for artifacts matching the given definitions.
    func scan(
        root: URL,
        definitions: [ArtifactDefinition],
        includeHidden: Bool = false,
        onProgress: @Sendable (String) -> Void,
        onFound: @Sendable (URL, ArtifactCategory, Int) -> Void
    ) -> [(url: URL, category: ArtifactCategory)] {
        cancelled = false
        var results: [(url: URL, category: ArtifactCategory)] = []
        let fm = FileManager.default

        // Build lookup structures
        let targetNames = ArtifactRegistry.allTargetNames(for: definitions)

        // Adjust skip list: remove any entries that are target directories
        // (e.g. .gradle should not be skipped if Gradle scanning is enabled)
        let skipDirs = Self.baseSkipDirs.subtracting(targetNames)

        var stack: [(URL, Int)] = [(root, 0)]
        let maxDepth = 8
        var lastProgressTime = Date.timeIntervalSinceReferenceDate

        while !stack.isEmpty {
            if cancelled { return results }

            let (dirURL, depth) = stack.removeLast()
            if depth > maxDepth { continue }

            let now = Date.timeIntervalSinceReferenceDate
            if now - lastProgressTime > 0.1 {
                onProgress(dirURL.path)
                lastProgressTime = now
            }

            guard let contents = try? fm.contentsOfDirectory(
                at: dirURL,
                includingPropertiesForKeys: [.isDirectoryKey, .isSymbolicLinkKey],
                options: [.skipsPackageDescendants]
            ) else { continue }

            for itemURL in contents {
                if cancelled { return results }

                let name = itemURL.lastPathComponent

                // Check if this directory name matches any artifact target
                if targetNames.contains(name) {
                    guard let values = try? itemURL.resourceValues(
                        forKeys: [.isDirectoryKey, .isSymbolicLinkKey]
                    ) else { continue }

                    if values.isDirectory == true && values.isSymbolicLink != true {
                        // Check which definition matches (including sibling detection)
                        if let def = ArtifactRegistry.matchDefinition(
                            dirName: name,
                            parentURL: dirURL,
                            definitions: definitions,
                            fm: fm
                        ) {
                            results.append((url: itemURL, category: def.category))
                            onFound(itemURL, def.category, results.count)
                        }
                    }
                    // Don't recurse into matched directories regardless of match
                    continue
                }

                // Only recurse into directories
                guard let values = try? itemURL.resourceValues(
                    forKeys: [.isDirectoryKey, .isSymbolicLinkKey]
                ) else { continue }

                guard values.isDirectory == true else { continue }
                guard values.isSymbolicLink != true else { continue }

                // Skip known unproductive directories
                if skipDirs.contains(name) { continue }

                // Skip hidden dirs unless opted in (but never skip artifact targets)
                if !includeHidden && name.hasPrefix(".") && !targetNames.contains(name) { continue }

                stack.append((itemURL, depth + 1))
            }
        }

        return results
    }

    /// Check system-level artifact paths for existence.
    /// Returns entries for each path that exists.
    /// For paths with expandSubdirectories, returns individual subdirectory entries.
    nonisolated func checkSystemArtifacts(
        categories: Set<ArtifactCategory>
    ) -> [(url: URL, category: ArtifactCategory)] {
        let fm = FileManager.default
        var results: [(url: URL, category: ArtifactCategory)] = []

        for sysArtifact in ArtifactRegistry.systemArtifacts(for: categories) {
            guard fm.fileExists(atPath: sysArtifact.url.path) else { continue }

            if sysArtifact.expandSubdirectories {
                // List subdirectories as individual entries
                guard let contents = try? fm.contentsOfDirectory(
                    at: sysArtifact.url,
                    includingPropertiesForKeys: [.isDirectoryKey],
                    options: [.skipsHiddenFiles]
                ) else {
                    // If we can't list, add the whole directory
                    results.append((url: sysArtifact.url, category: sysArtifact.category))
                    continue
                }

                for item in contents {
                    guard let values = try? item.resourceValues(forKeys: [.isDirectoryKey]),
                          values.isDirectory == true else { continue }
                    results.append((url: item, category: sysArtifact.category))
                }
            } else {
                results.append((url: sysArtifact.url, category: sysArtifact.category))
            }
        }

        return results
    }
}
